File: 1834.single-threaded-cpu.py
#
# @lc app=leetcode id=1834 lang=python3
#
# [1834] Single-Threaded CPU
#
from heapq import *
from typing import *

# @lc code=start
class Solution:
    # T: O(n * log(n))
    # S: O(n)
    def getOrder(self, tasks: List[List[int]]) -> List[int]:
        # Sort by enqueue time and then processing time. CPU must deal with tasks
        # in the order of the enqueue time.
        tasks = sorted([(start, length, i) for i, (start, length) in enumerate(tasks)])
        
        heap = []
        cur_time = 0
        res = [0] * len(tasks)
        i = 0

        for start, length, task_id in tasks:
            # If the CPU is idle and there are available tasks, 
            # the CPU will choose the one with the shortest 
            # processing time. If multiple tasks have the 
            # same shortest processing time, it will choose 
            # the task with the smallest index. As heapq is
            # min heap, element in the heap should be a tuple 
            # (proc_time, index, ...).
            while len(heap) and cur_time < start:
                # Perform a task
                next_length, next_id, next_start = heappop(heap)
                res[i] = next_id
                i += 1
                cur_time = max(cur_time, next_start) + next_length
            heappush(heap, (length, task_id, start))

        while len(heap):
            # Perform the remaining tasks
            res[i] = heappop(heap)[1]
            i += 1
        return res

# @lc code=end

File: 1834.single-threaded-cpu.java
/*
 * @lc app=leetcode id=1834 lang=java
 *
 * [1834] Single-Threaded CPU
 */
import java.util.*;

// @lc code=start
class Solution {
    // T: O(n * log(n))
    // S: O(n)
    public int[] getOrder(int[][] tasks) {
        final int n = tasks.length;
        int[][] allTasks = new int[n][3];
        for (int i = 0; i < n; ++i) {
            allTasks[i][0] = i;
            allTasks[i][1] = tasks[i][0];
            allTasks[i][2] = tasks[i][1];
        }

        // sort by enqueue time
        Arrays.sort(allTasks, (a, b) -> Integer.compare(a[1], b[1]));

        // Compare process time and then the original indices of the tasks
        PriorityQueue<int[]> pq = new PriorityQueue<int[]>((a, b) -> a[2] == b[2] ? Integer.compare(a[0], b[0]) : Integer.compare(a[2], b[2]));
        int curTime = 0;
        int[] res = new int[n];
        int taskIdx = 0, resIdx = 0;
        while (resIdx < n) {
            while (taskIdx < n && allTasks[taskIdx][1] <= curTime) {
                pq.offer(allTasks[taskIdx++]);
            }

            if (pq.isEmpty()) {
                // Idle after a task is done (next enqueue time > current finish time), 
                // process the next task available.
                curTime = allTasks[taskIdx][1];
                continue;
            }

            int[] selectedTask = pq.poll();
            res[resIdx++] = selectedTask[0];
            curTime += selectedTask[2];
        }
        return res;
    }
}
// @lc code=end

